native/dsterf.go

```diff
@@ -0,0 +1,292 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/lapack"
+)
+
+// Dsterf computes all eigenvalues of a symmetric tridiagonal matrix using the
+// Pal-Walker-Kahan variant of the QL or QR algorithm.
+//
+// d contains the diagonal elements of the tridiagonal on entry, and contains the
+// eigenvalues in ascending order on exit. d must have length at least n, or
+// Dsterf will panic.
+//
+// e contains the off-digonal elements of the tridiagonal on entry, and is
+// overwritten during the call to Dsterf.
+func (impl Implementation) Dsterf(n int, d, e []float64) (ok bool) {
+	if n < 0 {
+		panic(nLT0)
+	}
+	if n == 0 {
+		return true
+	}
+	if len(d) < n {
+		panic(badD)
+	}
+	if len(e) < n-1 {
+		panic(badE)
+	}
+
+	const (
+		none = 0 // no scaling
+		down = 1 // Scale the values below ssfmax threshhold
+		up   = 2 // Scale the values below ssfmin threshhold
+	)
+
+	// Determine the unit roundoff for this environment.
+	eps := dlamchE
+	eps2 := eps * eps
+	safmin := dlamchS
+	safmax := 1 / safmin
+	ssfmax := math.Sqrt(safmax) / 3
+	ssfmin := math.Sqrt(safmin) / eps2
+
+	// Compute the eigenvalues of the tridiagonal matrix.
+	maxit := 30
+	nmaxit := n * maxit
+	jtot := 0
+
+	l1 := 0
+
+	// TODO(btracey): Define these closer to use when gotos are removed.
+	var anorm, c, gamma, r, rte, s, sigma float64
+	var iscale, l, lend, lendsv, lsv, m int
+
+	// TOOD(btracey): Replace these goto statements with imperative flow control
+	// structures.
+Ten:
+	if l1 > n-1 {
+		goto OneSeventy
+	}
+	if l1 > 0 {
+		e[l1-1] = 0
+	}
+	for m = l1; m < n-1; m++ {
+		if math.Abs(e[m]) <= math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1]))*eps {
+			e[m] = 0
+			goto Thirty
+		}
+	}
+	m = n - 1
+
+Thirty:
+	l = l1
+	lsv = l
+	lend = m
+	lendsv = lend
+	l1 = m + 1
+	if lend == 0 {
+		goto Ten
+	}
+
+	// Scale submatrix in rows and columns l to lend.
+	anorm = impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+	iscale = none
+	if anorm == 0 {
+		goto Ten
+	}
+	if anorm > ssfmax {
+		iscale = down
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l, 1, e[l:], n)
+	} else if anorm < ssfmin {
+		iscale = up
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l, 1, e[l:], n)
+	}
+
+	for i, v := range e[l:lend] {
```

Comment:
Done. It can't be as you wrote there because el comes into scope. I put it in a block, but I can declare it at the top instead if you'd prefer.